src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -133,6 +133,30 @@ public static decimal Abs(decimal value)
             return decimal.Abs(value);
         }
 
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static double Abs(double value)
+        {
+            // Implementation based on https://git.musl-libc.org/cgit/musl/tree/src/math/fabs.c
+
+            const ulong signMask = 0x7FFFFFFFFFFFFFFF;
```

Comment:
In theory we could just do
```csharp
return Sse2.And(
    Vector128.CreateScalarUnsafe(x),
    Vector128.Create(0x7FFFFFFFFFFFFFFFUL).AsDouble()).ToScalar();
```
and get rid of `[Intrinsic]` but we won't be able to fold constant input

Reply:
Yup, I was comparing ryjit vs. clang.
```c
// C
#include <inttypes.h>

double fabs(double value)
{
    uint64_t u = *(uint64_t*)&value & 0x7FFFFFFFFFFFFFFF;
    return *(double*)&u;
}

float fabsf(float value)
{
    uint32_t u = *(uint32_t*)&value & 0x7FFFFFFF;
    return *(float*)&u;
}
```

```c#
// C#
static unsafe double fabs(double value)
{
    ulong u = *(ulong*)&value & 0x7FFFFFFFFFFFFFFF;
    return *(double*)&u;
}

/* OR without `unsafe`
static double Abs(double value)
{
    // Implementation based on https://git.musl-libc.org/cgit/musl/tree/src/math/fabs.c

    const ulong signMask = 0x7FFFFFFFFFFFFFFF;
    ulong raw = BitConverter.DoubleToUInt64Bits(value);

    return BitConverter.UInt64BitsToDouble(raw & signMask);
}
*/

static unsafe float fabsf(float value)
{
    uint u = *(uint*)&value & 0x7FFFFFFF;
    return *(float*)&u;
}

/* OR without `unsafe`
static float Abs(float value)
{
    // Implementation based on https://git.musl-libc.org/cgit/musl/tree/src/math/fabsf.c

    const uint signMask = 0x7FFFFFFF;
    uint raw = BitConverter.SingleToUInt32Bits(value);

    return BitConverter.UInt32BitsToSingle(raw & signMask);
}
*/
```

clang: deals with a single `vandps` instruction
```asm
.LCPI0_0:
        .quad   0x7fffffffffffffff              # double NaN
        .quad   0x7fffffffffffffff              # double NaN
fabs:                                   # @fabs
        vandps  xmm0, xmm0, xmmword ptr [rip + .LCPI0_0]
        ret
.LCPI1_0:
        .long   0x7fffffff                      # float NaN
fabsf:                                  # @fabsf
        vandps  xmm0, xmm0, dword ptr [rip + .LCPI1_0]{1to4}
        ret
```

ryujit: suffers from redundant moves
```asm
fabs(Double):
    vzeroupper
    vmovsd [rsp+8], xmm0
    mov rax, 0x7fffffffffffffff
    and rax, [rsp+8]
    vmovq xmm0, rax
    ret

fabsf(Single):
    vzeroupper
    vmovss [rsp+8], xmm0
    mov eax, [rsp+8]
    and eax, 0x7fffffff
    vmovd xmm0, eax
    ret
```

Directly using h/w intrinsic (`Sse.And(Vector128<float>,Vector128<float>)`) slightly improves the situation for float:

```c#
static float fabsf(float value)
{
    Vector128<float> vec = Vector128.CreateScalarUnsafe(value);
    Vector128<float> mask = Vector128.CreateScalarUnsafe((float)0x7FFFFFFF);

    return X86.Sse.And(vec, mask).ToScalar();
}
```
```asm
fabsf(Single)
    vzeroupper
    vmovss xmm0, [0x7ffb39050458]
    vandps xmm0, xmm1, xmm0
    ret
```

but still no way to skip extra move and `vzeroupper` (https://github.com/dotnet/runtime/issues/10484).

Reply:
I don't think vzeroupper is a problem - we emit it even today for `Math.Abs` intrinsic, more like that `vmovss` but it's fixed if you propagate `mask` into And ("contain"):
![image](https://user-images.githubusercontent.com/523221/149778316-053d7f36-a040-40b4-8a10-672a1cae2bdc.png)

Reply:
Pushed a change with hardware intrinsic. I can look into deleting `[Intrinsic]` and cleaning up JIT next.

Reply:
Your commit regresses constant input - jit won't be able to fold `Math.Abs(-3.14)` also your changes regress Mono since it doesn't support S.R.I intrinsics outside of LLVM mode.
Also, Inliner treats calls with [Intrinsic] specially